From db43c8249b965717d65ac41a54f7f039bde90d7b Mon Sep 17 00:00:00 2001
From: Sarai <shearthstone@habana.ai>
Date: Tue, 20 May 2025 13:15:31 +0300
Subject: [PATCH] [SW-229337] emb: build: pyhlml: fix generator

Use IntEnum instead of Enum to allow direct access of values.
Allow user-defined types to be used as types later on.
Delete unused imports.

Change-Id: I5c83edd3b5588aa221a5f3e2204f78c31878ca66
Signed-off-by: Sarai <shearthstone@habana.ai>
---

diff --git a/pyhlml/pyhlml/generator/custom_pycparser.py b/pyhlml/pyhlml/generator/custom_pycparser.py
index 5ea5e3a..19439de 100644
--- a/pyhlml/pyhlml/generator/custom_pycparser.py
+++ b/pyhlml/pyhlml/generator/custom_pycparser.py
@@ -3,7 +3,7 @@
 """
 
 from pycparser import c_ast
-from utils import is_namespace_node, get_c_type, add_to_ctypes_map
+from utils import is_namespace_node, get_c_type, add_to_ctypes_map, is_known_struct, add_to_known_struct_names
 
 class CustomVisitor(c_ast.NodeVisitor):
     def __init__(self, namespace: str):
@@ -101,13 +101,22 @@
                             field_type = f"ctypes.POINTER({base_type_name})"
 
                 # Handle regular types
-                elif isinstance(field.type, c_ast.TypeDecl) and hasattr(field.type.type, "names"):
-                    field_type = " ".join(field.type.type.names)
-                    field_type = get_c_type(field_type)
+                elif isinstance(field.type, c_ast.TypeDecl):
+                    decl_type = field.type.type
+                    if isinstance(decl_type, c_ast.IdentifierType):
+                        type_name = " ".join(decl_type.names)
+                        # Check if it's a known struct (user-defined)
+                        if is_known_struct(type_name):
+                            # prefix with 'c_' and remove trailing '_t'
+                            if type_name.endswith('_t'):
+                                type_name = type_name[:-2]
+                            field_type = f"c_{type_name}"
+                        else:
+                            field_type = get_c_type(type_name)
 
                 if field_type:
                     struct_fields.append((field.name, field_type))
-        add_to_ctypes_map(field.name)
+        add_to_known_struct_names(struct_name)
         self.elements.append((struct_name, struct_fields))
 
 
diff --git a/pyhlml/pyhlml/generator/main.py b/pyhlml/pyhlml/generator/main.py
index 6e1c5c1..1344cb4 100644
--- a/pyhlml/pyhlml/generator/main.py
+++ b/pyhlml/pyhlml/generator/main.py
@@ -50,7 +50,7 @@
 def generate_python_code(typedefs: str, enums: str, structs: str, defines: str) -> str:
     code = (
         "import ctypes\n"
-        "from enum import Enum\n\n"
+        "from enum import IntEnum\n\n"
         "# Typedefs\n\n"
         + "\n".join(typedefs)
         + "\n\n# Enums\n\n"
diff --git a/pyhlml/pyhlml/generator/utils.py b/pyhlml/pyhlml/generator/utils.py
index 442e3b0..895ee1c 100644
--- a/pyhlml/pyhlml/generator/utils.py
+++ b/pyhlml/pyhlml/generator/utils.py
@@ -3,9 +3,6 @@
 """
 
 import re
-import tokenize
-from io import StringIO
-
 
 # This map is populated by generate_enum_code and used by qualify_consts
 ENUM_VALUE_TO_CLASS_MAP = {}
@@ -25,6 +22,9 @@
     "float": "c_float"
 }
 
+# Populated by the StructVisitor.visit_Struct method
+KNOWN_STRUCT_NAMES = []
+
 def add_to_ctypes_map(name):
     # Add struct name to known C types
     C_TYPE_NAMES_TO_CTYPES[name] = f"c_{name}"
@@ -33,11 +33,21 @@
 def get_c_type(lookup_type):
     if lookup_type is None:
         return None
+    elif lookup_type in KNOWN_STRUCT_NAMES:
+        return f"c_{lookup_type}"
     try:
         c_type = C_TYPE_NAMES_TO_CTYPES[lookup_type]
     except KeyError:
         c_type = "c_void_p"  # Fallback
     return f"ctypes.{c_type}"
+
+def add_to_known_struct_names(name):
+    KNOWN_STRUCT_NAMES.append(name)
+    KNOWN_STRUCT_NAMES.append(f"{name}_t")
+
+def is_known_struct(struct_name):
+    return struct_name in KNOWN_STRUCT_NAMES
+
 
 def is_namespace_node(node_name, namespace):
     return re.match(fr"^{namespace}", node_name, re.IGNORECASE)
@@ -67,7 +77,7 @@
 
 def generate_enum_code(enum_name: str, enum_members: dict) -> str:
     output = []
-    output.append(f"class {enum_name.upper()}(Enum):")
+    output.append(f"class {enum_name.upper()}(IntEnum):")
     last_value = -1
     for element_name, element_value in enum_members.items():
         if element_value is None:
